File: AWS/iac-microbenchmark/ec2/go/ecc256_verify/main.go
// main.go
package main

import (
	"context"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"os"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/config"
	"github.com/aws/aws-sdk-go-v2/service/kms"
	"github.com/aws/aws-sdk-go-v2/service/kms/types"
)

// ECC256Request represents the input message and signature
type ECC256Request struct {
	Message   string `json:"message"`
	Signature string `json:"signature"`
}

// ECC256Response represents the validation response
type ECC256Response struct {
	Valid bool `json:"valid"`
}

func main() {
	if len(os.Args) < 2 {
		fmt.Println(`{"error": "Missing JSON input"}`)
		return
	}

	// Get the JSON input from command line argument
	requestJSONRaw := os.Args[1]

	var messageStruct ECC256Request
	if err := json.Unmarshal([]byte(requestJSONRaw), &messageStruct); err != nil || messageStruct.Message == "" || messageStruct.Signature == "" {
		fmt.Println(`{"error": "Invalid request body, missing message or signature"}`)
		return
	}

	// Create a top-level context
	ctx := context.Background()

	// Load AWS configuration
	cfg, err := config.LoadDefaultConfig(ctx)
	if err != nil {
		fmt.Println(`{"error": "Failed to load AWS config"}`)
		return
	}

	// Initialize KMS client
	kmsClient := kms.NewFromConfig(cfg)

	// Get the KMS key ARN from environment variables
	eccKmsKeyID := os.Getenv("ECC256_KMS_KEY_ARN")
	if eccKmsKeyID == "" {
		fmt.Println(`{"error": "ECC 256 KMS key ARN not set"}`)
		return
	}

	// Call the function to verify the signature using KMS
	verificationResponse, err := kmsClientVerifyMessage(ctx, kmsClient, eccKmsKeyID, messageStruct.Message, messageStruct.Signature)
	if err != nil {
		fmt.Printf(`{"error": "Error verifying message: %v"}`, err)
		return
	}

	// Prepare the response body
	responseBody, _ := json.Marshal(verificationResponse)

	// Print the response
	fmt.Println(string(responseBody))
}

// kmsClientVerifyMessage verifies the message signature using ECC256 and AWS KMS
func kmsClientVerifyMessage(ctx context.Context, kmsClient *kms.Client, keyID string, message string, signatureB64 string) (*ECC256Response, error) {
	// Decode the base64-encoded signature
	signature, err := base64.StdEncoding.DecodeString(signatureB64)
	if err != nil {
		return nil, fmt.Errorf("failed to decode signature: %w", err)
	}

	// Perform the verification with KMS
	verifyOutput, err := kmsClient.Verify(ctx, &kms.VerifyInput{
		KeyId:            aws.String(keyID),
		Message:          []byte(message),
		MessageType:      types.MessageTypeRaw,
		Signature:        signature,
		SigningAlgorithm: types.SigningAlgorithmSpecEcdsaSha256,
	})
	if err != nil {
		return nil, fmt.Errorf("failed to verify message: %w", err)
	}

	// Prepare the response
	return &ECC256Response{
		Valid: verifyOutput.SignatureValid,
	}, nil
}

File: AWS/iac-microbenchmark/ec2/rust/ecc256_verify/src/main.rs
use aws_sdk_kms::{self as kms, primitives::Blob, Client};
use base64::{decode, encode};
use serde::{Deserialize, Serialize};
use std::env;
use std::error::Error;

#[derive(Serialize, Deserialize, Debug)]
struct ECC256Request {
    message: String,
    signature: String,
}

#[derive(Serialize, Deserialize, Debug)]
struct ECC256Response {
    valid: bool,
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn Error>> {
    // Get the KMS key ARN from environment variables
    let ecc_kms_key_id =
        env::var("ECC256_KMS_KEY_ARN").map_err(|_| "ECC 256 KMS key ARN not set?")?;

    // Get JSON input from command line
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        return Err("Usage: ./rustexecutable '{\"message\":\"your_message_here\", \"signature\":\"your_signature_here\"}'".into());
    }

    let body_str = &args[1];

    // Parse the input JSON
    let message_struct: ECC256Request = serde_json::from_str(body_str)
        .map_err(|_| "Invalid Request Body, missing message or signature.")?;

    // Initialize the KMS client
    let kms_client = kms::Client::new(&aws_config::load_from_env().await);

    let message = message_struct.message;
    let signature = message_struct.signature;

    // Attempt to verify the message using KMS
    match kms_client_verify_message(&kms_client, &ecc_kms_key_id, &message, &signature).await {
        Ok(verify_response) => {
            // Serialize the verification response to a JSON string
            let str_verify_response = serde_json::to_string(&verify_response)?;
            println!("{}", str_verify_response);
            Ok(())
        }
        Err(e) => Err(format!("Error verifying message: {}", e).into()),
    }
}

async fn kms_client_verify_message(
    kms_client: &Client,
    key_id: &str,
    message: &str,
    signature_b64: &str,
) -> Result<ECC256Response, Box<dyn Error>> {
    // Decode the base64-encoded signature
    let signature = decode(signature_b64).map_err(|_| "Failed to decode signature ECC256")?;

    let signature_blob = Blob::new(signature);
    let message_blob = Blob::new(message.as_bytes());

    // Perform the verification
    let verify_output = kms_client
        .verify()
        .key_id(key_id)
        .message(message_blob)
        .message_type(aws_sdk_kms::types::MessageType::Raw)
        .signature(signature_blob)
        .signing_algorithm(aws_sdk_kms::types::SigningAlgorithmSpec::EcdsaSha256)
        .send()
        .await?;

    let is_valid = verify_output.signature_valid;

    let verify_response = ECC256Response { valid: is_valid };

    Ok(verify_response)
}
